community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/PhysicalLogFile.java

```diff
@@ -207,31 +261,45 @@ public static PhysicalLogVersionedStoreChannel openForVersion( PhysicalLogFiles
                     fileToOpen.getCanonicalPath() ) );
         }
 
-        StoreChannel rawChannel;
+        StoreChannel rawChannel = null;
+        boolean successful = false;
         try
         {
-            rawChannel = fileSystem.open( fileToOpen, "rw" );
+            rawChannel = fileSystem.open( fileToOpen, write ? "rw" : "r" );
+
+            ByteBuffer buffer = ByteBuffer.allocate( LOG_HEADER_SIZE );
+            LogHeader header = readLogHeader( buffer, rawChannel, true, fileToOpen );
+            assert header != null && header.logVersion == version;
+            PhysicalLogVersionedStoreChannel result =
+                    new PhysicalLogVersionedStoreChannel( rawChannel, version, header.logFormatVersion );
+            successful = true;
+            return result;
         }
         catch ( FileNotFoundException cause )
         {
             throw Exceptions.withCause( new FileNotFoundException( String.format( "File could not be opened [%s]",
                     fileToOpen.getCanonicalPath() ) ), cause );
         }
-
-        ByteBuffer buffer = ByteBuffer.allocate( LOG_HEADER_SIZE );
-        LogHeader header = readLogHeader( buffer, rawChannel, true );
-        assert header != null && header.logVersion == version;
-
-        return new PhysicalLogVersionedStoreChannel( rawChannel, version, header.logFormatVersion );
+        finally
+        {
+            if (    // we managed to open the file
+                    rawChannel != null &&
+                    // but we didn't successfully do the rest, f.ex. reading the header
+                    !successful )
+            {
+                // then close the channel
+                rawChannel.close();
+                // ... and the exception causing us to not be successful will be thrown
+            }
```

Comment:
True, although I still need the `successful` flag to know if the `rawChannel` should be closed or not in the `finally` block, won't I?